### PR TITLE
Feature/customizable auth providers epmdj 10709

### PR DESCRIPTION
**Pull request by iryabov**

### CODE REVIEW

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        get("/oauth/callback") {
+            val oauthPrincipal: OAuthAccessTokenResponse.OAuth2? = call.principal()
+            if (oauthPrincipal == null) {
+                call.unauthorizedError()
+                return@get
+            }
+
+            val userInfo = getUserInfo(httpClient, oauthConfig.userInfoUrl, oauthPrincipal.accessToken)
+            if (userInfo == null) {
+                call.unauthorizedError()
```
Comment (RomanDavlyatshin) [inline admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt:117]: __comment 1__
[RomanDavlyatshin](https://github.com/RomanDavlyatshin) [5 days ago](https://github.com/Drill4J/admin/pull/340#discussion_r1389279980)
can't we just do throw UnauthorizedError() and use centralized error handling / status pages to handle that?

__comment 2__
@[iryabov](https://github.com/iryabov) iryabov [13 minutes ago](https://github.com/Drill4J/admin/pull/340#discussion_r1394112004)
Oauth2 routes are attached outside of the main routes and common error handlers for them may not work


__comment 3__

> are attached outside of the main routes

Hmm, I would say none [of the status pages ](https://github.com/Drill4J/admin/blob/feature/oauth-sessions-EPMDJ-10707/admin-core/src/main/kotlin/com/epam/drill/admin/DrillApplication.kt#L53) are coupled with routes down-the-line, aren't they?

I had an impression Ktor simply "places" all status pages as a sort-of-a middleware, that is applied to all routes/requests handlers

Have you checked if that works?

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthConfig.kt [hunk 2]
```diff
@@ -16,11 +16,55 @@
...
+    val issuer: String
+        get() = oauth2.property("issuer").getString()
+
+    val scopes: List<String>
+        get() = oauth2.propertyOrNull("scopes")?.getString()
+            ?.split(",")?.map { it.trim() }?.filter { it.isNotEmpty() }
+            ?: listOf()
+
+    val uiRootUrl: String
+        get() =  ui.propertyOrNull("rootUrl")?.getString() ?: "http://localhost:9090"
```
Comment (RomanDavlyatshin) [inline admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthConfig.kt]: no defaults provided for other values - do we really need to declare these here?

File: admin-core/src/main/kotlin/com/epam/drill/admin/DrillApplication.kt [hunk 3]
```diff
@@ -37,13 +45,64 @@ import io.ktor.routing.*
...
+@Suppress("unused")
+fun Application.moduleWithOAuth() {
+    installPlugins()
+    initDB()
+    di {
+        import(drillAdminDIModule)
+        import(oauthDIModule)
+    }
+    install(Authentication) {
+        configureOAuthAuthentication(closestDI())
```
Comment (RomanDavlyatshin) [inline admin-core/src/main/kotlin/com/epam/drill/admin/DrillApplication.kt:111]: _tbd_ closestDI

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+
+val oauthDIModule = DI.Module("oauth") {
+    configureOAuthDI()
+}
+
+fun DI.Builder.configureOAuthDI() {
+    bind<HttpClient>("oauthHttpClient") with singleton { HttpClient(Apache) }
+    bind<OAuthConfig>() with singleton { OAuthConfig(di) }
+    bind<JwkProvider>() with singleton { buildJwkProvider(instance()) }
```
Comment (RomanDavlyatshin) [inline admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt:55]: Both `OAuthConfig` and `JwkProvider` are only used at `Authentication.Configuration.configureOAuthAuthentication` function.
Are they added via kodein just to allow mocking and keep the DI approach consistent?

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+                return@get
+            }
+
+            val userInfo = getUserInfo(httpClient, oauthConfig.userInfoUrl, oauthPrincipal.accessToken)
+            if (userInfo == null) {
+                call.unauthorizedError()
+                return@get
+            }
+
+            call.response.cookies.append(
```
Comment (RomanDavlyatshin) [inline admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt:121]: I'm not sure anymore whether returning JWT in cookie will work
Ktor seems to be hardcoded to obtain token from header (see `io.ktor.auth.jwt.JWTAuthKt`, function `public fun Authentication.Configuration.jwt(`, line `val token = provider.authHeader(call)`)

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+            call.respondRedirect(oauthConfig.uiRootUrl)
+        }
+    }
+}
+
+private suspend fun getUserInfo(
+    httpClient: HttpClient,
+    userInfoUrl: String,
+    accessToken: String
+): JsonElement? {
```
Comment (RomanDavlyatshin) [inline admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt]: Stylistic preference - I suggest rewriting it to:
```
): JsonElement? =
    httpClient
        .get<HttpResponse>(userInfoUrl) {
            headers {
                append(HttpHeaders.Authorization, "Bearer $accessToken")
            }
        }
        .takeIf { it.status.isSuccess() }
        ?.let { it.receive<String>() }
        ?.let { Json.parseToJsonElement(it) }
```

Less words to read, no throwaway `val`-s, looks a bit more idiomatic to Kotlin style

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    val response =
+        httpClient.get<HttpResponse>(userInfoUrl) {
+            headers {
+                append(HttpHeaders.Authorization, "Bearer $accessToken")
```
Comment (RomanDavlyatshin) [inline admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt]: Is `Authorization` header and `Bearer $token` a unified approach to call API for all oauth2 implementations?

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+        }
+    return response.status.takeIf { it.isSuccess() }?.let { _ ->
+        val stringBody = response.receive<String>()
+        val jsonBody = Json.parseToJsonElement(stringBody)
+        jsonBody
+    }
+}
+
+private fun JsonElement.toPrincipal(): User {
+    val jsonObject = this.jsonObject
```
Comment (RomanDavlyatshin) [inline admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt]: Unnecessary `val` declaration - jsonObject is already available from `this` context under the same name

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+    return User(
+        username = jsonObject.getValue("preferred_username").jsonPrimitive.content,
+        role = jsonObject["roles"]?.jsonArray?.map { it.jsonPrimitive.content }.let { findRole(it ?: emptyList()) }
```
Comment (RomanDavlyatshin) [inline admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt]: Passing emptyList() seems odd - we _know_ find role won't match anything if input list parameter is empty

We might as well change `findRole` to accept nullable list
```
private fun JsonElement.toPrincipal(): User = User(
    username = jsonObject.getValue("preferred_username").jsonPrimitive.content,
    role = jsonObject["roles"]
        ?.jsonArray
        ?.map { it.jsonPrimitive.content }
        .let { findRole(it) }
)

private fun findRole(roleNames: List<String>?): Role = roleNames
    ?.takeIf { !it.isNullOrEmpty() } // this also avoids calling map and find for an empty list
    ?.distinct()
    ?.map { it.lowercase() }
    ?.let { roleNamesList ->
        Role.values().find { role ->
            roleNamesList.contains(role.name.lowercase())
        }
    }
    ?: Role.UNDEFINED
```

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+private fun JWTCredential.toPrincipal() = User(
+    username = this["preferred_username"].toString(),
+    role = findRole(this.payload.getClaim("realm_access").asMap()["roles"] as List<String>)
+)
+
+private fun findRole(roleNames: List<String>): Role {
+    return Role.values().find { role ->
+        roleNames.toSet().map { it.lowercase() }.contains(role.name.lowercase())
```
Comment (RomanDavlyatshin) [inline admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt]: I think `toSet` can be replaced with `distinct` for clarity
but we better test that

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+    } ?: Role.UNDEFINED
+}
+
+private fun buildJwkProvider(oauthConfig: OAuthConfig): JwkProvider {
+    return JwkProviderBuilder(URL(oauthConfig.jwkSetUrl))
+        .cached(10, 24, TimeUnit.HOURS)
+        .rateLimited(10, 1, TimeUnit.MINUTES)
```
Comment (RomanDavlyatshin) [inline admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt:174]: we wanted to make these values adjustable with `application.conf`

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/SecurityConfig.kt
Comment (RomanDavlyatshin) [inline admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/SecurityConfig.kt:1]: Looks like `SecurityConfig` became `OAuthModule` -> if that's the case - I kindly remind to use `git mv` to rename files

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/SimpleAuthModule.kt [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+        realm = "Access to the http(s) services"
+        validate {
+            authService.signIn(LoginPayload(username = it.name, password = it.password)).toPrincipal()
+        }
+    }
+}
+
+private fun DI.Builder.userServicesConfig() {
+    bind<UserAuthenticationService>() with singleton {
+        val app: Application = instance()
```
Comment (RomanDavlyatshin) [inline admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/SimpleAuthModule.kt]: I suggest rewriting to 
```
    bind<UserAuthenticationService>() with singleton {
        UserAuthenticationServiceImpl(
            userRepository = instance(),
            passwordService = instance()
        )
        .let {
            when (instance<Application>().userRepoType) {
                UserRepoType.DB -> TransactionalUserAuthenticationService(it)
                else -> it
            }
        }
    }
```

Alternatively it could be expressed in a more imperative style:
```
    bind<UserAuthenticationService>() with singleton {
        val service = UserAuthenticationServiceImpl(
            userRepository = instance(),
            passwordService = instance()
        )
        if (instance<Application>().userRepoType == UserRepoType.DB)
            TransactionalUserAuthenticationService(service)
        else
            service
    }
```
although, I prefer the first variant

TL;DR Please do not use `if-else` inside `.let` 😄

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/SimpleAuthModule.kt [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+        ).let { service ->
+            if (app.userRepoType == UserRepoType.DB)
+                TransactionalUserAuthenticationService(service)
+            else
+                service
+        }
+    }
+    bind<UserManagementService>() with singleton {
+        val app: Application = instance()
+        UserManagementServiceImpl(
```
Comment (RomanDavlyatshin) [inline admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/SimpleAuthModule.kt:118]: same-as-above refactoring is applicable here

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/SimpleAuthModule.kt [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+    bind<PasswordGenerator>() with singleton { PasswordGeneratorImpl(config = instance()) }
+    bind<PasswordValidator>() with singleton { PasswordValidatorImpl(config = instance()) }
+    bind<PasswordService>() with singleton { PasswordServiceImpl(instance(), instance()) }
+}
+
+private fun DI.Builder.userRepositoriesConfig() {
+    bind<UserRepository>() with singleton {
+        val app: Application = instance()
+        logger.info { "The user repository type is ${app.userRepoType}" }
+        when (app.userRepoType) {
```
Comment (RomanDavlyatshin) [inline admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/SimpleAuthModule.kt:137]: `        when (instance<Application>().userRepoType) {`

File: admin-core/src/main/kotlin/com/epam/drill/admin/endpoints/agent/AgentHandler.kt [hunk 15]
```diff
@@ -35,119 +33,121 @@ import io.ktor.http.cio.websocket.*
...
-                }
-                agentSession.createWsLoop(
-                    agentInfo,
-                    call.request.headers[ContentEncoding] == "deflate"
-                )
-            }
+            else -> FrameType.BINARY
+        }
+        val agentSession = AgentWsSession(
+            this@agentWebsocket,
```
Comment (RomanDavlyatshin) [inline admin-core/src/main/kotlin/com/epam/drill/admin/endpoints/agent/AgentHandler.kt]: `this@agentWebsocket` 😮 _tbd_

File: admin-core/src/main/kotlin/com/epam/drill/admin/endpoints/plugin/DrillPluginWs.kt [hunk 16]
```diff
@@ -122,3 +123,9 @@ class DrillPluginWs(override val di: DI) : DIAware {
         copy(buildVersion = agentManager.buildVersionByAgentId(agentId))
     } else this
 }
+
+fun Routing.pluginWebSocketRoute() {
+    val handler by closestDI().instance<DrillPluginWs>()
+
```
Comment (RomanDavlyatshin) [inline admin-core/src/main/kotlin/com/epam/drill/admin/endpoints/plugin/DrillPluginWs.kt]: unnecessary vertical whitespace

File: admin-core/src/main/kotlin/com/epam/drill/admin/endpoints/plugin/PluginDispatcher.kt [hunk 17]
```diff
@@ -434,6 +435,12 @@ internal class PluginDispatcher(override val di: DI) : DIAware {
     }
 }
 
+fun Routing.pluginDispatcherRoutes() {
+    val pluginDispatcher by closestDI().instance<PluginDispatcher>()
+
```
Comment (RomanDavlyatshin) [inline admin-core/src/main/kotlin/com/epam/drill/admin/endpoints/plugin/PluginDispatcher.kt]: remove whitespace

File: admin-core/src/main/kotlin/com/epam/drill/admin/kodein/KodeinModules.kt [hunk 18]
```diff
@@ -23,84 +23,71 @@ import com.epam.drill.admin.cache.impl.*
...
-        bind<PluginDispatcher>() with eagerSingleton { PluginDispatcher(di) }
-        bind<VersionEndpoints>() with eagerSingleton { VersionEndpoints(di) }
-        bind<GroupHandler>() with eagerSingleton { GroupHandler(di) }
-        bind<AgentHandler>() with eagerSingleton { AgentHandler(di) }
-        bind<NotificationEndpoints>() with eagerSingleton { NotificationEndpoints(di) }
-        bind<RequestValidator>() with eagerSingleton { RequestValidator(di) }
+val wsHandler = DI.Module("wsHandler") {
+    bind<DrillPluginWs>() with eagerSingleton { DrillPluginWs(di) }
+
+    bind<TopicResolver>() with eagerSingleton {
```
Comment (RomanDavlyatshin) [inline admin-core/src/main/kotlin/com/epam/drill/admin/kodein/KodeinModules.kt:75]: this can be oneliner

File: admin-core/src/main/resources/application.conf [hunk 19]
```diff
@@ -6,7 +6,7 @@ ktor {
         sslPort = ${?SSL_PORT}
     }
     application {
-        modules = [com.epam.drill.admin.DrillApplicationKt.module]
+        modules = [com.epam.drill.admin.DrillApplicationKt.moduleWithSimpleAuth]
```
Comment (RomanDavlyatshin) [inline admin-core/src/main/resources/application.conf]: lets add env variable to pick either simpleAuth or OAuth

File: admin-auth/src/test/kotlin/com/epam/drill/admin/auth/TestUtils.kt [hunk 20]
```diff
@@ -71,3 +79,17 @@ fun <M, T> wheneverBlocking(mock: M, methodCall: suspend M.() -> T): OngoingStub
     return runBlocking { whenever(mock.methodCall()) }
 }
 
+suspend fun HttpRequestData.formData(): Map<String, String> {
```
Comment (RomanDavlyatshin) [inline admin-auth/src/test/kotlin/com/epam/drill/admin/auth/TestUtils.kt]: Stylistic preference: avoid using throwaway `val`-s and `for...in`
```
suspend fun HttpRequestData.formData(): Map<String, String> = String(this.body.toByteArray())
    .split("&")
    .map { it.split("=") }
    .associate { (key, value) ->
        key to withContext(Dispatchers.IO) {
            URLDecoder.decode(value, "UTF-8")
        }
    }
```

File: admin-auth/src/test/kotlin/com/epam/drill/admin/auth/OAuthModuleTest.kt [hunk 21]
```diff
@@ -0,0 +1,245 @@
...
+    fun `oauth login request must be redirected to oauth2 authorize url`() {
+        withTestApplication({
+            withOAuthTestEnvironment()
+            withTestOAuthModule()
+        }) {
+            with(handleRequest(HttpMethod.Get, "/oauth/login")) {
+                assertEquals(HttpStatusCode.Found, response.status())
+
+                val url = URL(response.headers[HttpHeaders.Location])
+                val queryParams = url.query?.split("&")?.associate {
```
Comment (RomanDavlyatshin) [inline admin-auth/src/test/kotlin/com/epam/drill/admin/auth/OAuthModuleTest.kt]: move parsing code to `parseQueryParams` function (either lambda right here or move it to the test utils) to improve readability

File: admin-auth/src/test/kotlin/com/epam/drill/admin/auth/OAuthModuleTest.kt [hunk 22]
```diff
@@ -0,0 +1,245 @@
...
+    @Test
+    fun `oauth login request must be redirected to oauth2 authorize url`() {
+        withTestApplication({
+            withOAuthTestEnvironment()
+            withTestOAuthModule()
+        }) {
+            with(handleRequest(HttpMethod.Get, "/oauth/login")) {
+                assertEquals(HttpStatusCode.Found, response.status())
+
+                val url = URL(response.headers[HttpHeaders.Location])
```
Comment (RomanDavlyatshin) [inline admin-auth/src/test/kotlin/com/epam/drill/admin/auth/OAuthModuleTest.kt]: rename `url` to `redirectedUrl`  or `locationHeaderValue`?

**Review by RomanDavlyatshin — CHANGES_REQUESTED** [review]

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/SecurityConfig.kt
Comment (iryabov) [reply admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/SecurityConfig.kt:1]: Thanks for reminding. In this case the idea to turn SecurityConfig into an OAuthModule was not implemented in one change

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/SimpleAuthModule.kt (hunk 14 above)
Comment (iryabov) [reply admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/SimpleAuthModule.kt:137]: app is used twice: ${app.userRepoType}

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt (hunk 1 above)
Comment (iryabov) [reply admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt:117]: Will decide later

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt (hunk 4 above)
Comment (iryabov) [reply admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt:55]: Yes

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt (hunk 5 above)
Comment (iryabov) [reply admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt:121]: I use authHeader function

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt (hunk 7 above)
Comment (iryabov) [reply admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt]: It works for Keycloak

File: admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt (hunk 11 above)
Comment (iryabov) [reply admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/OAuthModule.kt:174]: It is caching of keycloak certificates

File: admin-core/src/main/kotlin/com/epam/drill/admin/DrillApplication.kt [hunk 23]
```diff
@@ -54,9 +54,11 @@ private val logger = KotlinLogging.logger {}
 
 @Suppress("unused")
 fun Application.module() {
-    when (authType) {
-        AuthType.SIMPLE -> moduleWithSimpleAuth()
-        AuthType.OAUTH -> moduleWithOAuth()
+    when (authType.lowercase()) {
+        "simple" -> moduleWithSimpleAuth()
```
Comment (RomanDavlyatshin) [inline admin-core/src/main/kotlin/com/epam/drill/admin/DrillApplication.kt]: 1. Swap "simple" and "oauth" string literals for `AuthType.SIMPLE.name` and ` AuthType.OAUTH.name` to avoid string duplication
2. Should we better call it "OAUTH2", just to eliminate possible confusion?

File: admin-core/src/main/kotlin/com/epam/drill/admin/DrillApplication.kt [hunk 24]
```diff
@@ -54,9 +54,11 @@ private val logger = KotlinLogging.logger {}
...
+        "oauth" -> moduleWithOAuth()
+        else -> throw IllegalArgumentException("Unknown auth type: $authType. " +
+                "Please set the env variable DRILL_AUTH_TYPE to either 'simple' or 'oauth'.")
```
Comment (RomanDavlyatshin) [inline admin-core/src/main/kotlin/com/epam/drill/admin/DrillApplication.kt]: get possible values straight from the enum type (to make sure it message wont get outdated if values change): 
```
    "...to either ${AuthType.values().joinToString(separator = "\", \"", prefix = "\"", postfix = "\"") { it.name }}"
```

File: admin-core/src/main/kotlin/com/epam/drill/admin/DrillApplication.kt [hunk 25]
```diff
@@ -112,14 +114,19 @@ fun Application.moduleWithOAuth() {
...
+    }
 }
 
-private val Application.authType: AuthType
+private val Application.authType: String
     get() = environment.config
         .propertyOrNull("drill.auth.type")
-        ?.getString()?.let { AuthType.valueOf(it.uppercase()) }
-        ?: AuthType.SIMPLE
+        ?.getString() ?: "simple"
```
Comment (RomanDavlyatshin) [inline admin-core/src/main/kotlin/com/epam/drill/admin/DrillApplication.kt]: Use `AuthType.SIMPLE.name`

**Review by RomanDavlyatshin — CHANGES_REQUESTED** [review]

**Review by RomanDavlyatshin — APPROVED** [review]
